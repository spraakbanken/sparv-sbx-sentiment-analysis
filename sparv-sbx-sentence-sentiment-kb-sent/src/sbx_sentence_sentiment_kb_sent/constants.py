"""Constants used in this project."""

PROJECT_NAME: str = "sbx_sentence_sentiment_kb_sent"
