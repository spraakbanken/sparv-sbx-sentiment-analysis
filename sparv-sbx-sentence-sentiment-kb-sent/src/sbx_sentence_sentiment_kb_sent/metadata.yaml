id: sbx-swe-sentiment-transformers-kblab_robust_swedish_sentiment_multiclass
short_description:
    swe: "Denna analys har bytt namn: https://spraakbanken.gu.se/analyser/sbx-swe-sentiment_sentence-transformers-kblab_robust_swedish_sentiment_multiclass"
    eng: "This analysis has changed name: https://spraakbanken.gu.se/en/analyses/sbx-swe-sentiment_sentence-transformers-kblab_robust_swedish_sentiment_multiclass"
description:
    swe: |-
        Denna analys har bytt namn:
        https://spraakbanken.gu.se/analyser/sbx-swe-sentiment_sentence-transformers-kblab_robust_swedish_sentiment_multiclass
    eng: |-
        This analysis has changed name:
        https://spraakbanken.gu.se/en/analyses/sbx-swe-sentiment_sentence-transformers-kblab_robust_swedish_sentiment_multiclass
task: sentiment analysis
language_codes:
    - swe
keywords:
    - transformers
    - kb-lab
annotations:
    - <sentence>:sbx_sentence_sentiment_kb_sent.sentence-sentiment--kb-sent
example_output: |-
    ```xml
    <sentence sentence-sentiment--kb-sent="|NEUTRAL:0.946|">
    <token pos="JJ">Stora</token>
    <token pos="NN">regnmängder</token>
    <token pos="VB">väntas</token>
    <token pos="PP">under</token>
    <token pos="NN">måndagen</token>
    <token pos="KN">och</token>
    <token pos="PM">SMHI</token>
    <token pos="VB">har</token>
    <token pos="VB">utfärdat</token>
    <token pos="DT">en</token>
    <token pos="JJ">gul</token>
    <token pos="NN">varning</token>
    <token pos="PP">för</token>
    <token pos="PC">skyfallsliknande</token>
    <token pos="NN">regn</token>
    <token pos="PP">över</token>
    <token pos="JJ">stora</token>
    <token pos="NN">delar</token>
    <token pos="PP">av</token>
    <token pos="NN">landets</token>
    <token pos="JJ">södra</token>
    <token pos="JJ">halva</token>
    <token pos="MAD">.</token>
    </sentence>
    ```
tools:
    - name: "transformers"
      url: "https://github.com/huggingface/transformers"
      license: "Apache-2.0 license"
models:
    - name: "KBLab/robust-swedish-sentiment-multiclass"
      url: "https://huggingface.co/KBLab/robust-swedish-sentiment-multiclass"
      description: |-
          KBLab presents a robust, multi-label sentiment classifier trained on Swedish texts.
          The model is robust in the sense that it is trained on multiple datasets of different
          text types and allows labeling of neutral as well positive and negative texts.
      license: "Apache-2.0 license"
other_references:
    - "https://kb-labb.github.io/posts/2023-06-16-a-robust-multi-label-sentiment-classifier-for-swedish/"
created: 2024-05-28
updated: 2025-10-24
