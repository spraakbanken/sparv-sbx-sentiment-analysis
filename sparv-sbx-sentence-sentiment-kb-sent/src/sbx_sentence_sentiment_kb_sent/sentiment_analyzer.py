"""Sentiment analyzer."""

from collections import defaultdict
from typing import Dict, List, Optional, Union

from sparv import api as sparv_api  # type: ignore [import-untyped]
from transformers import (  # type: ignore [import-untyped]
    AutoModelForSequenceClassification,
    AutoTokenizer,
    MegatronBertForSequenceClassification,
    PreTrainedTokenizerFast,
    pipeline,
)

logger = sparv_api.get_logger(__name__)

TOKENIZER_NAME = "KBLab/megatron-bert-large-swedish-cased-165k"
TOKENIZER_REVISION = "90c57ab49e27b820bd85308a488409dfea25600d"
MODEL_NAME = "KBLab/robust-swedish-sentiment-multiclass"
MODEL_REVISION = "b0ec32dca56aa6182a6955c8f12129bbcbc7fdbd"

TOK_SEP = " "
MAX_LENGTH: int = 700


class SentimentAnalyzer:
    """Sentiment analyzer."""

    def __init__(
        self,
        *,
        tokenizer: Optional[PreTrainedTokenizerFast] = None,
        model: Optional[MegatronBertForSequenceClassification] = None,
        num_decimals: int = 3,
    ) -> None:
        """Create a SentimentAnalyzer using the given tokenizer and model.

        The given number of num_decimals works both as rounding and cut-off.

        Args:
            tokenizer (PreTrainedTokenizerFast): the tokenizer to use
            model (MegatronBertForSequenceClassification): the model to use
            num_decimals (int): number of decimals to use (defaults to 3)
        """
        self.tokenizer = self._default_tokenizer() if tokenizer is None else tokenizer
        self.model = self._default_model() if model is None else model
        self.num_decimals = num_decimals
        self.classifier = pipeline(
            "sentiment-analysis", model=self.model, tokenizer=self.tokenizer
        )

    @classmethod
    def _default_tokenizer(cls) -> PreTrainedTokenizerFast:
        return AutoTokenizer.from_pretrained(TOKENIZER_NAME, revision=TOKENIZER_REVISION)

    @classmethod
    def _default_model(cls) -> MegatronBertForSequenceClassification:
        return AutoModelForSequenceClassification.from_pretrained(
            MODEL_NAME, revision=MODEL_REVISION
        )

    @classmethod
    def default(cls) -> "SentimentAnalyzer":
        """Create a SentimentAnalyzer with default tokenizer and model.

        Returns:
            SentimentAnalyzer: the create SentimentAnalyzer
        """
        tokenizer = cls._default_tokenizer()
        model = cls._default_model()
        return cls(model=model, tokenizer=tokenizer)

    def analyze_sentence(self, text: List[str]) -> Optional[str]:
        """Analyze a sentence.

        Args:
            text (Iterable[str]): the text to analyze

        Returns:
            List[Optional[str]]: the sentence annotations.
        """
        total_length = sum(len(t) for t in text) + len(text) - 1
        logger.debug("analyzed text length=%d", total_length)
        if total_length > MAX_LENGTH:
            logger.warning(
                "Long sentence (%d chars), splitting and combining results", total_length
            )
            classifications = self._analyze_in_chunks(text)
        else:
            sentence = TOK_SEP.join(text)
            logger.debug("analyzing '%s'", sentence)
            classifications = self.classifier(sentence)
        logger.debug("classifications = %s", classifications)
        collect_label_and_score = ((clss["label"], clss["score"]) for clss in classifications)
        score_format, score_pred = SCORE_FORMAT_AND_PREDICATE[self.num_decimals]

        format_scores = (
            (label, score_format.format(score)) for label, score in collect_label_and_score
        )
        filter_out_zero_scores = (
            (label, score) for label, score in format_scores if not score_pred(score)
        )
        classification_str = "|".join(
            f"{label}:{score}" for label, score in filter_out_zero_scores
        )
        return f"|{classification_str}|" if classification_str else "|"

    def _analyze_in_chunks(self, text: List[str]) -> List[Dict[str, Union[str, float]]]:
        classifications_list = []
        start_i = 0
        curr_length = 0
        for t_i, t in enumerate(text):
            if len(t) + curr_length > MAX_LENGTH:
                clss = self.classifier(" ".join(text[start_i:t_i]))
                classifications_list.append(clss)
                start_i = t_i
            else:
                curr_length += len(t) + 1
        classifications_dict = defaultdict(list)
        for clsss in classifications_list:
            for clss in clsss:
                classifications_dict[clss["label"]].append(clss["score"])

        return [
            {"label": label, "score": sum(scores) / len(scores)}
            for label, scores in classifications_dict.items()
        ]


SCORE_FORMAT_AND_PREDICATE = {
    1: ("{:.1f}", lambda s: s.startswith("0") and s.endswith(".0")),
    2: ("{:.2f}", lambda s: s.startswith("0") and s.endswith(".00")),
    3: ("{:.3f}", lambda s: s.startswith("0") and s.endswith(".000")),
    4: ("{:.4f}", lambda s: s.startswith("0") and s.endswith(".0000")),
    5: ("{:.5f}", lambda s: s.startswith("0") and s.endswith(".00000")),
    6: ("{:.6f}", lambda s: s.startswith("0") and s.endswith(".000000")),
    7: ("{:.7f}", lambda s: s.startswith("0") and s.endswith(".0000000")),
    8: ("{:.8f}", lambda s: s.startswith("0") and s.endswith(".00000000")),
    9: ("{:.9f}", lambda s: s.startswith("0") and s.endswith(".000000000")),
    10: ("{:.10f}", lambda s: s.startswith("0") and s.endswith(".0000000000")),
}
