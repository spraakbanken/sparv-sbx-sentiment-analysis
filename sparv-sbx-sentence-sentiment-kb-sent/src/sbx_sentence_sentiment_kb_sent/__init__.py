"""Sparv plugin for annotating sentences with sentiment analysis."""

from sparv import api as sparv_api  # type: ignore [import-untyped]

from sbx_sentence_sentiment_kb_sent.annotations import annotate_sentence_sentiment
from sbx_sentence_sentiment_kb_sent.constants import PROJECT_NAME

__all__ = ["annotate_sentence_sentiment"]

__description__ = "Annotate sentence with sentiment analysis."
__version__ = "0.2.0"

__config__ = [
    sparv_api.Config(
        f"{PROJECT_NAME}.num_decimals",
        description="The number of decimals to round the score to",
        default=3,
    ),
]
