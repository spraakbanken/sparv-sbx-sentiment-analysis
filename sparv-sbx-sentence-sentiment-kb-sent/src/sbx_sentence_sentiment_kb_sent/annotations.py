"""Sparv annotator for sentiment analysis on sentences."""

from sparv import api as sparv_api  # type: ignore [import-untyped]

from sbx_sentence_sentiment_kb_sent.constants import PROJECT_NAME
from sbx_sentence_sentiment_kb_sent.sentiment_analyzer import SentimentAnalyzer

logger = sparv_api.get_logger(__name__)


@sparv_api.annotator("Sentiment analysis of sentences", language=["swe"])
def annotate_sentence_sentiment(
    out_sentence_sentiment: sparv_api.Output = sparv_api.Output(
        f"<sentence>:{PROJECT_NAME}.sentence-sentiment--kb-sent",
        # cls="sbx_sentence_sentiment_kb_sent",
        description="Sentiment analysis of sentence with KBLab/robust-swedish-sentiment-multiclass",  # noqa: E501
    ),
    word: sparv_api.Annotation = sparv_api.Annotation("<token:word>"),
    sentence: sparv_api.Annotation = sparv_api.Annotation("<sentence>"),
    num_decimals_str: str = sparv_api.Config(f"{PROJECT_NAME}.num_decimals"),
) -> None:
    """Sentiment analysis of sentence with KBLab/robust-swedish-sentiment-multiclass."""
    try:
        num_decimals = int(num_decimals_str)
    except ValueError as exc:
        raise sparv_api.SparvErrorMessage(
            f"'{PROJECT_NAME}.num_decimals' must contain an 'int' got: '{num_decimals_str}'"
        ) from exc
    sentences, _orphans = sentence.get_children(word)
    token_word = list(word.read())
    out_sentence_sentiment_annotation = sentence.create_empty_attribute()

    analyzer = SentimentAnalyzer(num_decimals=num_decimals)

    logger.progress(total=len(sentences))  # type: ignore
    for sent_i, sent in enumerate(sentences):
        sent_to_tag = [token_word[token_index] for token_index in sent]
        out_sentence_sentiment_annotation[sent_i] = analyzer.analyze_sentence(sent_to_tag)
        logger.debug("annotation[%d]=%s", sent_i, out_sentence_sentiment_annotation[sent_i])

    out_sentence_sentiment.write(out_sentence_sentiment_annotation)
