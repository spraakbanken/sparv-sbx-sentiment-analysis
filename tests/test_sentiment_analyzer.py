import pytest
from sbx_sentence_sentiment_kb_sent.sentiment_analyzer import SentimentAnalyzer


@pytest.fixture(name="sentiment_analyzer", scope="session")
def fixture_sentiment_analyzer() -> SentimentAnalyzer:
    return SentimentAnalyzer.default()


def test_neutral_text(sentiment_analyzer: SentimentAnalyzer) -> None:
    text = [
        "Vi",
        "hoppas",
        "och",
        "tror",
        "att",
        "detta",
        "också",
        "snabbt",
        "ska",
        "kunna",
        "komma",
        "på",
        "plats",
        ",",
        "säger",
        "Garborg",
        ".",
    ]

    actual = sentiment_analyzer.analyze_sentence(text)

    actual = remove_scores(actual)
    expected = "|NEUTRAL|"

    assert actual == expected


def remove_scores(actual: str | None) -> str | None:
    """Remove scores."""
    if not actual:
        return actual
    return "|".join(x.split(":")[0] for x in actual.split("|"))


def test_issue_10(sentiment_analyzer: SentimentAnalyzer, snapshot) -> None:  # noqa: ANN001
    text = [
        "Frisinnade",
        "landsföreningen,",
        "som",
        "på",
        "allt",
        "sätt",
        "vill",
        "verka",
        "till",
        "främjande",
        "af",
        "ett",
        "sådant",
        "resultat,",
        "anser,",
        "att",
        "den",
        "förestående",
        "valkampanjen,",
        "förutom",
        "kring",
        "den",
        "politiska",
        "rösträttsfrågan",
        "såsom",
        "den",
        "gifna",
        "hufvudfrågan,",
        "bör",
        "koncentrera",
        "sig",
        "särskildt",
        "kring",
        "följande",
        "uppgifter:\n\natt",
        "till",
        "Andra",
        "kammaren",
        "må",
        "sändas",
        "män,",
        "hvilka",
        "-",
        "särskildt",
        "med",
        "hänsyn",
        "till",
        "statsbudgetens",
        "oerhörda",
        "stegring",
        "-",
        "vilja",
        "häfda",
        "den",
        "sunda",
        "sparsamhetens",
        "grundsatser",
        "i",
        "vår",
        "statshushållning;\n\natt",
        "till",
        "Andra",
        "kammaren",
        "sändas",
        "män,",
        "hvilka",
        "-",
        "sedan",
        "riksdagen",
        "antagit",
        "en",
        "progressiv",
        "inkomstskatt",
        "i",
        "förening",
        "med",
        "obligatorisk",
        "själfdeklaration",
        "-",
        "vilja",
        "verka",
        "för",
        "bevillningsförordningens",
        "omarbetande",
        "i",
        "enlighet",
        "med\ndessa",
        "sålunda",
        "erkända",
        "grundsatser",
        "och",
        "därvid",
        "jämväl",
        "beakta,",
        "att",
        "taxeringsförfarandet",
        "äfven",
        "för",
        "icke",
        "deklarationsskyldige",
        "måtte",
        "blifva",
        "fullt",
        "betryggande;\n\natt",
        "till",
        "Andra",
        "kammaren",
        "sändas",
        "män,",
        "hvilka",
        "äro",
        "beslutna",
        "att",
        "verka",
        "för",
        "en",
        "effektiv",
        "begränsning",
        "af",
        "den",
        "kommunala",
        "röstskalan;\n\natt",
        "till",
        "Andra",
        "kammaren",
        "sändas",
        "män,",
        "hvilka",
        "-",
        "såsom",
        "en",
        "nödvändig",
        "konsekvens",
        "af",
        "1901",
        "års",
        "beslut",
        "i",
        "försvarsfrågan",
        "-",
        "vilja",
        "verka",
        "för",
        "en",
        "grundlig",
        "omarbetning",
        "af",
        "krigslagarne",
        "i",
        "tidsenlig",
        "och",
        "human",
        "riktning;\n\natt",
        "till",
        "Andra",
        "kammaren",
        "sändas",
        "män,",
        "hvilka",
        "äro",
        "beslutna",
        "att",
        "verka",
        "för",
        "folkskolans",
        "sunda",
        "utveckling",
        "och",
        "närmast",
        "för",
        "att",
        "äfven",
        "folkskolorna",
        "i",
        "-likhet",
        "med",
        "de",
        "allmänna",
        "läroverkan",
        "måtte",
        "erhålla",
        "sin",
        "särskilda",
        "öfverstyrelse;\n\natt",
        "till",
        "Andra",
        "kammaren",
        "sändas",
        "män,",
        "som",
        "vilja",
        "arbeta",
        "mot",
        "öfverhandtagande",
        "af",
        "bolagens",
        "jordvälde",
        "och",
        "för",
        "vidmakthållande",
        "och",
        "stärkande",
        "af",
        "en",
        "själfägande",
        "jordbrukande",
        "befolkning",
        "samt",
        "mot",
        "bolagsmakten",
        "i",
        "den",
        "kommunala",
        "styrelsen;\n\natt",
        "till",
        "Andra",
        "kammaren",
        "sändas",
        "män,",
        "hvilka",
        "fullt",
        "behjärta",
        "arbetarefrågans",
        "utomordentliga",
        "betydelse",
        "och",
        "särskildt",
        "såsom",
        "närmaste",
        "mål",
        "vilja",
        "verka",
        "för",
        "upprättande",
        "af",
        "ett",
        "statsdepartement",
        "för",
        "arbetarelagstiftning",
        "och",
        "därmed",
        "sammanhängande",
        "ärenden",
        "anordnande",
        "af",
        "opartisk",
        "medling",
        "i",
        "arbetstvister",
        "samt",
        "invaliditets-",
        "och",
        "ålderdomspensionering,",
        "och",
        "hvilka",
        "bestämdt",
        "motsätta",
        "sig",
        "all",
        "ensidig",
        "strafflagstiftning",
        "såsom",
        "botemedel",
        "mot",
        "sociala",
        "missförhållanden;\n\natt",
        "till",
        "Andra",
        "kammaren",
        "sändas",
        "män,",
        "hvilka",
        "i",
        "frågor",
        "sådana",
        "som",
        "näringslifvets",
        "höjande,",
        "främjande",
        "af",
        "det",
        "mindre",
        "jordbruket,",
        "utveckling",
        "af",
        "egna",
        "hensrörelsen,",
        "ny",
        "arrendelagstiftning,",
        "humanare",
        "fattigvårdslagstiftning,",
        "äfvenson",
        "tidsenlig",
        "lönereglering",
        "för",
        "de",
        "s.k.",
        "lägre",
        "statstjänarne,",
        "vilja",
        "nedlägga",
        "ett",
        "allvarligt",
        "och",
        "ihärdigt",
        "arbete",
        "för",
        "att",
        "få",
        "till",
        "stånd",
        "goda",
        "och",
        "gagnande",
        "resultat;\n\nsamt",
        "slutligen",
        "att",
        "till",
        "Andra",
        "kammaren",
        "sändas",
        "män,",
        "hvilka",
        "på",
        "nykterhetslagstiftningens",
        "viktiga.",
    ]

    actual = sentiment_analyzer.analyze_sentence(text)

    assert actual == snapshot
