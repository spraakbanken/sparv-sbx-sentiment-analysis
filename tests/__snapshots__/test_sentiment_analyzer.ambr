# serializer version: 1
# name: test_issue_10
  '|NEUTRAL:0.931|POSITIVE:0.700|NEGATIVE:0.743|'
# ---
